spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/core/MqttClientManager.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2002-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.mqtt.core;
+
+
+import java.util.Optional;
+
+import org.eclipse.paho.client.mqttv3.IMqttAsyncClient;
+import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
+import org.eclipse.paho.client.mqttv3.MqttException;
+
+/**
+ * Basic interface for managing clients.
+ * <p>Manages clients and connections as well as client callbacks.
+ *
+ * @author Artem Vozhdayenko
+ *
+ * @since 6.0
+ */
+public interface MqttClientManager {
```

Comment:
I didn't look into the problem in details, but my Idea was to have a direct `IMqttAsyncClient` injection into the target channel adapter.
It is already up to end-user to create that client and configure it properly.
The `clientId` would be out of channel adapter scope.
So, we would still have an option be be able to create a client per channel adapter as it is right now.
And another option is to rely on the externally configured client.

Can you elaborate, please, the reasoning behind this additional abstraction?

Reply:
My initial design was based on comment you have left in the original issue:

> So, for every new subscriber we can have its own listening context while keeping the rest of MqttCallback in the central place, e.g. MqttClientManager.
> Probably something similar what we have so far with the StompSessionManager or ClientWebSocketContainer.
> As far as I'm concerned I see that on publishing side the MqttClient doesn't have a state, but just gives us a DeliveryToken to track publication.

Maybe I misunderstood the idea?..


Regarding the approach when the client is injected by the user: in this case, it will be the responsibility of the user to manage client lifecycle? (Now in case of connection lost, the client is disconnected and closed and the new instance is created)
I am not sure if reopening the same client will work.. maybe this is something we need to check :-) 
I guess the responsibility of client manager is to maintain clients and link them with client ids. Something like a repository. This is done just for the usage convenience, so that a user does not bother with clients creation and management.

Reply:
I have looked through the paho client API docs and write a simple tests to verify: the client can be disconnected and then the same instance can be re-used. The only one method which prevents client to be re-used is `close()`.

```java
	@Test
	void testClientReuse() throws Exception {
		MqttAsyncClient mqttClient = new MqttAsyncClient(MosquittoContainerTest.mqttUrl(), "test-client-id");

		mqttClient.connect().waitForCompletion(1000);

		mqttClient.publish("test", new MqttMessage(new byte[] {1, 2, 3,}));

		mqttClient.disconnect().waitForCompletion(1000);

		mqttClient.connect().waitForCompletion(1000);

		assertThatNoException().isThrownBy(() -> mqttClient.publish("test", new MqttMessage(new byte[] {1, 2, 3})));
	}
```

So, theoretically we could have client injected in the Adapter/Handler instances, just as you said. But I am not sure about this `close()` method. It is already used in the current, 'production' implementation. Was it used because of some reasons? [it is called in case of exceptions / connection loss]
Maybe some issues with reusing the same instance...

Reply:
**UPD**
Just found that in case of "external" client injection, we should not disconnect from the "inner" code as it will break some other places where this client is used.
So, `onStop()` should just stop some loops/checks etc., maybe unsubscribe from topics, but not change the client state...

Reply:
All your investigations and conclusions are correct. My bad: I forgot what I suggested in that issue 😄 .

So, the idea is to have a manager component for a single client. Implement the lifecycle you have discovered and disconnect the lifecycle control in the channel adapters if such a manager is injected. Or maybe the client it manages. Perhaps there is really no reason to let channel adapters to know about the manager.

A callback you mention probably better to have internally in the manager and don't expose anything from the channel adapter for the external client. Since such a callback must be single and controlled from the manager. 
The only callback a channel adapter must provide is the one for a listener or publish API.
The lifecycle of a client and reconnect logic must be implemented in that manager. 
I'd prefer an existing logic to be present in the channel adapters if we create clients as it is now.
Probably only difference that listener and publish callbacks could be reused since their purpose remains the same independently of the client management.

Does it make sense?

Thank you!

Reply:
So, my current vision is:
- we do not change anything in the existing adapters/handlers, current clients/client factories should be still in place for backwards compatibility;
- we need to have new adapters/handler implementations where we're injecting an external client (can call it `sharedClient`) from the client manager;
- Those more lightweight _adapters_ do not contain reconnect logic and only provide respective callbacks to the `clientManager` (that is, a callback for failed/successful delivery, for a new message);
- Those more lightweight _message handlers_ should basically have only one difference: no client connect operation performed and close/disconnect on `stop()` method as well;
- Client manager keeps a single client instance manageable using lifecycle methods so that the correct usage should be:
```java
@Autowired ClientManager clientManager1;
@Autowired ClientManager clientManager2;
clientManager1.start();
clientManager2.start();
client1 = clientManager1.getClient();
client2 = clientManager1.getClient();
adapter1 = new SharedClientAdapter(client1);
adapter2 = new SharedClientAdapter(client2);
handler1 = new SharedClientHandler(client1);
handler2 = new SharedClientHandler(client2);
...
adapter1.stop();
...
clientManager.stop();
```

Are we on the same page here? Or do you want to keep a *single* implementation and choose the strategy based on which constructor has been used, whether the users injects the client (new behavior) or uses the default ones (old behavior). I think in this case the classes could become more complex and messy, lots of additional if's. Maybe that could be solved using some sort of strategy pattern: a single facade class but different implementation delegates under the hood, but I am not sure if this would be a good approach at the end...

Reply:
Sorry, but no: we are not on the same page.

I still think that we need to keep and just extend the logic of the current channel adapters. 
Yes, the logic will be a bit complicated, but that's already internals of those components: the end-user would still expect from us a single entry point even if it would bring more choices of the configuration.

So, it looks like a `ClientManager` abstraction is only what we need to introduce and according to the existing `AbstractMqttMessageHandler` or `AbstractMqttMessageDrivenChannelAdapter`, we don't have a choice unless inject exactly this `ClientManager` to keep in as abstract as possible.
From here it sounds like `ClientManager` should be generic and have impls for MQTT v3 & v5.
We call `getClient()` in the target impl and switch to the respective logic. 
The common callback and reconnect logic is handled in the `ClientManager`.
We change the logic to the `public IMqttToken subscribe(String topicFilter, int qos, Object userContext, MqttActionListener callback)` in the `AbstractMqttMessageDrivenChannelAdapter` impls for any use-case instead of a global one as it is right now.
We change the logic to the `publish(String topic, MqttMessage message, Object userContext, MqttActionListener callback)` in the `AbstractMqttMessageHandler` impls.

My point is to not override a common callback provided by the `ClientManager` impl.
We may have some simple one by default: kinda copy (or mimic) of the existing `MqttCallback` impl for either channel adapter. And then we may look into some way to let end-user to configure their own `MqttCallback`.

Reply:
You know I came up with another thought.

Since you worry about mess in the logic of the current channel adapters, we may consider to create an internal `ClientManager` based on the provided `url` & `clientId`.
This way the client lifecycle logic will still be done only in a single place - that `ClientManager` impl, independently of the way we provide the client.
Right, the channel adapter will still control the lifecycle for that internal manager because it is not controlled externally.

I'm OK either way: just an idea in case you are pursuing perfection 😉 .
Only what I'm against of is a new set of channel adapters: this is an option for end-user to configure existing channel adapter some other way and nail the problem of a "single client per application".

Reply:
I also considered such a way but there is one potential issue. I am not sure about the probability, but if we have several adapters / handlers each using the same client manager and we want to shut down only one of them - then the client will be closed and unusable for other consumers.

Reply:
@artembilan can you please take a look at the new changes - I have pushed some sketches in this PR (with the new interface and one example of adapter change for v3)

Reply:
If the manager is external, then adapter must not control the client lifecycle . So shouting down adapter must not effect others.